### DISCUSSION TITLE
[SSR] How to render Lit Web Components in Nuxt with SSR?

### DISCUSSION THREAD

**Discussion by abdonrd** [lit/lit#2259] [General]
We have a few projects built in [Nuxt v2](https://nuxtjs.org/). Also trying the new [Nuxt v3](https://v3.nuxtjs.org/).
And we want to import some Web Components built with Lit v2.

It works if we disable the Nuxt SSR. But we want to be able to keep it active.
v2: https://nuxtjs.org/docs/configuration-glossary/configuration-ssr/
v3: https://v3.nuxtjs.org/guide/deploy/static-hosting#client-side-only-rendering

Has anyone already worked with it? Any ideas? Thanks!

**Comment by justinfagnani** [comment]
What breaks?

**Reply by abdonrd** [reply]
In Nuxt v3, with just import a Lit component:

```
window is not defined
  at file://./node_modules/@lit/reactive-element/css-tag.js:6:9  
  at ModuleJob.run (internal/modules/esm/module_job.js:170:25)  
  at async Loader.import (internal/modules/esm/loader.js:178:24)  
  at async __instantiateModule__ (file://./.nuxt/dist/server/server.mjs:1830:3)
```

And it works if you add `ssr: false` in the `nuxt.config.ts`.

Copied to here for visibility: https://github.com/lit/lit/discussions/2259#discussioncomment-1524771

**Reply by abdonrd** [reply]
In Nuxt v2, with just import a Lit component:

>Must use import to load ES Module: `/Users/abdonrd/Desktop/nuxt2-lit/node_modules/@qiskit/banner/index.js` require() of ES modules is not supported. require() of `/Users/abdonrd/Desktop/nuxt2-lit/node_modules/@qiskit/banner/index.js` from `/Users/abdonrd/Desktop/nuxt2-lit/node_modules/vue-server-renderer/build.dev.js` is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules. Instead rename index.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from `/Users/abdonrd/Desktop/nuxt2-lit/node_modules/@qiskit/banner/package.json`.

But then you add this to the `nuxt.config.ts` and fix that ES Module issue:

```javascript
export default {
  build: {
    standalone: true,
  }
}
```

That's because of https://github.com/nuxt/nuxt.js/issues/9223#issuecomment-840747064.

Now you have the same error as Nuxt v3:

```
window is not defined
```

And also it works if you add `ssr: false` in the `nuxt.config.ts`.

Copied to here for visibility: https://github.com/lit/lit/discussions/2259#discussioncomment-1524767

**Reply by justinfagnani** [reply]
Can you exclude certain modules from loading in the SSR environment?

**Reply by abdonrd** [reply]
I don't know. Googling now...

**Reply by abdonrd** [reply]
Anyway... Excluding Lit will not generate FOUC?

**Reply by justinfagnani** [reply]
Without lit-ssr actually integrated into into Nuxt's - which will presumably take some work - there's no other option. The Lit components will only render client-side. The key to minimizing it would be to only exclude the Lit components from the server build and not the client build.

**Reply by abdonrd** [reply]
> Can you exclude certain modules from loading in the SSR environment?

With Nuxt v2, seems like we can load the components as a plugin with the `mode: 'client'`:

```javascript
export default {
  plugins: [
    { src: '~/plugins/both-sides.js' },
    { src: '~/plugins/client-only.js', mode: 'client' }, // only on client side
    { src: '~/plugins/server-only.js', mode: 'server' } // only on server side
  ]
}
```

https://nuxtjs.org/docs/directory-structure/plugins/#object-syntax

It works, but the FOUC is bad. 😕

**Comment by abdonrd** [comment]
In Nuxt v2, with just import a Lit component:

>Must use import to load ES Module: `/Users/abdonrd/Desktop/nuxt2-lit/node_modules/@qiskit/banner/index.js` require() of ES modules is not supported. require() of `/Users/abdonrd/Desktop/nuxt2-lit/node_modules/@qiskit/banner/index.js` from `/Users/abdonrd/Desktop/nuxt2-lit/node_modules/vue-server-renderer/build.dev.js` is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules. Instead rename index.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from `/Users/abdonrd/Desktop/nuxt2-lit/node_modules/@qiskit/banner/package.json`.

But then you add this to the `nuxt.config.ts` and fix that ES Module issue:

```javascript
export default {
  build: {
    standalone: true,
  }
}
```

That's because of https://github.com/nuxt/nuxt.js/issues/9223#issuecomment-840747064.

---

Now the error is:

```
window is not defined
```

And it works if you add `ssr: false` in the `nuxt.config.ts`.

**Comment by abdonrd** [comment]
In Nuxt v3, with just import a Lit component:

```
window is not defined
  at file://./node_modules/@lit/reactive-element/css-tag.js:6:9  
  at ModuleJob.run (internal/modules/esm/module_job.js:170:25)  
  at async Loader.import (internal/modules/esm/loader.js:178:24)  
  at async __instantiateModule__ (file://./.nuxt/dist/server/server.mjs:1830:3)
```

And it works if you add `ssr: false` in the `nuxt.config.ts`.

**Comment by eliasmohsengrenvald** [comment]
In order to avoid `window is undefined` you would need to do the following as a very first one before all other imports:
`import '@lit-labs/ssr/lib/install-global-dom-shim.js'`

**Comment by eliasmohsengrenvald** [comment]
Also have a look at my `Nuxt` config under this sample repo, it also fixes the other issue with `Must use import to load ES Module`, you would have to add `lit` to `build>transpile`:
https://github.com/mohsenasfia/nuxt2

I managed to fix all those issues but in the end I'm getting another error `cannot find module 'parse5’` highlighted here #2559

**Comment by aomarks** [comment]
Hi @abdonrd!

We've just published some pre-releases for Lit which includes https://github.com/lit/lit/pull/3156. This change means that Lit should now be importable directly from Node, without crashing, without the need to first load the global DOM shim.

You should be able to upgrade with:

```sh
npm i lit@^2.3.0-next.1
```

This should allow you to keep Nuxt SSR enabled, but still be able to import Lit components.

Note this won't enable actual SSR of Lit components out-of-the-box, but it should at least let you client-side render them without integration problems on the server side.

Also note that this does _not_ automatically globally define any DOM APIs like the global DOM shim does, apart from `customElements` which is shimmed as a no-op. So if you have any calls like `window.addEventListener` running at module initialization, that will still error.

If you have a chance to check out this pre-release, that would be great, and please comment here to let us know how that works for you!

**Reply by abdonrd** [reply]
Thanks for all the work and info, @aomarks! 👏

It seems that it works better.

1. I just create a new Nuxt v3 project: `npx nuxi init nuxt-app`
2. Update a Lit component library: `npm i @qiskit/web-components`
3. Update the `app.vue` to add a Lit component:
    ```vue
    <template>
      <div>
        <qiskit-banner>HOLA</qiskit-banner>
        <NuxtWelcome />
      </div>
    </template>

    <script setup lang="ts">
    import "@qiskit/web-components/banner";
    </script>
    ```
4. The terminal show this error:
    ```
    [Vue warn]: Failed to resolve component: qiskit-banner
    If this is a native custom element, make sure to exclude it from component resolution via compilerOptions.isCustomElement.
    [nitro] [dev] [unhandledRejection] TypeError: Invalid value used as weak map key
        at WeakMap.set (<anonymous>)
        at normalizePropsOptions (/Users/abdonrd/Desktop/nuxt-app-v3.0.0-rc.6/node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:3966:11)
        at createComponentInstance (/Users/abdonrd/Desktop/nuxt-app-v3.0.0-rc.6/node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:6957:23)
        at renderComponentVNode (/Users/abdonrd/Desktop/nuxt-app-v3.0.0-rc.6/node_modules/@vue/server-renderer/dist/server-renderer.cjs.js:171:22)
        at Module.ssrRenderComponent (/Users/abdonrd/Desktop/nuxt-app-v3.0.0-rc.6/node_modules/@vue/server-renderer/dist/server-renderer.cjs.js:608:12)
        at _sfc_ssrRender (file:///Users/abdonrd/Desktop/nuxt-app-v3.0.0-rc.6/.nuxt/dist/server/server.mjs:3348:31)
        at renderComponentSubTree (/Users/abdonrd/Desktop/nuxt-app-v3.0.0-rc.6/node_modules/@vue/server-renderer/dist/server-renderer.cjs.js:253:13)
        at renderComponentVNode (/Users/abdonrd/Desktop/nuxt-app-v3.0.0-rc.6/node_modules/@vue/server-renderer/dist/server-renderer.cjs.js:188:16)
        at Module.ssrRenderComponent (/Users/abdonrd/Desktop/nuxt-app-v3.0.0-rc.6/node_modules/@vue/server-renderer/dist/server-renderer.cjs.js:608:12)
        at default (file:///Users/abdonrd/Desktop/nuxt-app-v3.0.0-rc.6/.nuxt/dist/server/server.mjs:2607:37)
    ```
5. I fixed it adding this to the `nuxt.config.ts`:
    ```typescript
    vue: {
      compilerOptions: {
        isCustomElement: (tag: string) => tag.includes('-'),
      },
    },
    ```
6. The terminal show this error:
    ```
    ReferenceError: window is not defined
        at file:///Users/abdonrd/Desktop/nuxt-app-v3.0.0-rc.6/node_modules/@lit/reactive-element/css-tag.js:6:9
        at ModuleJob.run (node:internal/modules/esm/module_job:197:25)
        at async Promise.all (index 0)
        at async ESMLoader.import (node:internal/modules/esm/loader:337:24)
        at async __instantiateModule__ (file:///Users/abdonrd/Desktop/nuxt-app-v3.0.0-rc.6/.nuxt/dist/server/server.mjs:3635:3)
    ```
7. I fixed it adding this to the `package.json` and redoing the `npm install`:
    ```json
    "overrides": {
      "lit": "2.3.0-next.1",
      "lit-element": "3.2.3-next.1",
      "lit-html": "2.3.0-next.1"
    }
    ```
8. It works!
    <img width="1195" alt="Screen Shot 2022-07-26 at 11 37 09" src="https://user-images.githubusercontent.com/1007051/180975277-9e9d034f-c266-4c9e-b3d1-09a1e8967de2.png">

The FOUC is bad, but it works!

**Reply by abdonrd** [reply]
> Note this won't enable actual SSR of Lit components out-of-the-box, but it should at least let you client-side render them without integration problems on the server side.

@aomarks do you have an example or more info regarding how to SSR the Lit components in Nuxt v3? Thanks!

**Reply by songkeys** [reply]
I still got `ReferenceError: HTMLElement is not defined` in Next.js. it points to this line which was originally located at `node_modules/lit/node_modules/@lit/reactive-element/reactive-element.js`

![image](https://user-images.githubusercontent.com/22665058/181710480-9ca4cd61-8e3c-4aa3-a686-9c0cb36e413c.png)


edit:
I tried to import [`browser-env`](https://www.npmjs.com/package/browser-env) to solve this. And now it's `ReferenceError: customElements is not defined` in this file: `node_modules/lit/node_modules/@lit/reactive-element/decorators/custom-element.js`

**Reply by aomarks** [reply]
@songkeys Are you using the `next-1` versions?

**Reply by songkeys** [reply]
Yes! I'm on the `next-1` version using the script mentioned ahove in package.json:

```json
"overrides": {
    "lit": "2.3.0-next.1",
    "lit-element": "3.2.3-next.1",
    "lit-html": "2.3.0-next.1"
  },
```

I'm bundling my lit-element as a npm library for other users. Could it be that I *bundled* the lit libraries together with only one possible environment? Are they using different entry files for different environments (node / browser)?

**Reply by aomarks** [reply]
Yes, that would explain the issue @Songkeys. The new Node-compatibility feature works by having a different build in the `node/` folder, which the `node` export condition is configured to point at in the `package.json`. So if you are bundling, you will instead be getting the `default` export condition, which is the browser build.

In general, we don't recommend bundling when creating a redistributable library. We recommend just keeping bare module specifiers, and letting the end user decide if and how they want to bundle. Allowing users to take advantage of this new Node compatibility build of Lit is one advantage of that. Another advantage is that it prevents duplication of Lit, if say your end user is also using Lit, or is using another library that uses Lit.

See also https://github.com/lit/lit/issues/2524#issuecomment-1199930523 where another user hit the same issue. Rollup and Webpack both support specifying the import condition, so in theory you can make a node-specific bundled build that follows the `node` export condition. But again, I would suggest just not bundling in this case.

**Comment by augustjk** [comment]
@abdonrd If you're still working with Nuxt, I wanted to highlight the work being done by @prashantpalikhe and @steveworkman here: https://github.com/prashantpalikhe/nuxt-ssr-lit

**Reply by abdonrd** [reply]
Thanks for the info, @augustjk!

**Reply by abdonrd** [reply]
@prashantpalikhe @steveworkman it seems that a big change about SSR is coming in the next version: https://github.com/lit/lit/pull/3554

**Reply by steveworkman** [reply]
Yes, we are very excited! I contributed SSR's breaking change and the Nuxt plugin will need it for full 3.0 stable compatibility. We will update the Nuxt plugin as soon as it's out

**Reply by abdonrd** [reply]
Oh! That sounds great! Thanks!
